cysystemd/async_reader.py

```diff
@@ -133,30 +126,53 @@ def skip_next(self, skip):
     def previous(self, skip=0):
         return self._exec(self.__reader.previous, skip)
 
-    async def skip_previous(self, skip):
-        return await self._exec(self.__reader.skip_previous, skip)
+    def skip_previous(self, skip):
+        return self._exec(self.__reader.skip_previous, skip)
+
+    def add_filter(self, rule):
+        return self._exec(self.__reader.add_filter, rule)
 
-    async def add_filter(self, rule):
-        return await self._exec(self.__reader.add_filter, rule)
+    def clear_filter(self):
+        return self._exec(self.__reader.clear_filter)
 
-    async def clear_filter(self):
-        return await self._exec(self.__reader.clear_filter)
+    def next(self, skip=0):
+        return self._exec(self.__reader.next, skip)
+
+    def __aiter__(self):
+        if self.__iterator and not self.__iterator.closed:
+            raise RuntimeError('Previous iterator still active')
+
+        self.__iterator = AsyncReaderIterator(
+            loop=self._loop, executor=self._executor, reader=self.__reader
+        )
+        return self.__iterator
+
+
+class AsyncReaderIterator(Base, AsyncIterator):
+    __slots__ = "reader", "queue", "closed"
+
+    QUEUE_SIZE = 1024
+
+    def __init__(self, *, reader, loop, executor):
+        super().__init__(loop=loop, executor=executor)
+        self.reader = reader
+        self.queue = asyncio.Queue(self.QUEUE_SIZE)
+        self._loop.create_task(self._exec(self._reader))
+        self.closed = False
 
-    async def __aiter__(self):
-        def read_entries():
-            for item in self.__reader:
-                asyncio.run_coroutine_threadsafe(queue.put(item), self._loop)
+    def _reader(self):
+        for item in self.reader:
+            asyncio.run_coroutine_threadsafe(self.queue.put(item), self._loop)
 
-            asyncio.run_coroutine_threadsafe(queue.put(None), self._loop)
+        self.closed = True
+        asyncio.run_coroutine_threadsafe(self.queue.put(None), self._loop)
```

Comment:
The "tail" of created and not awaited coroutines in the loop will schedule not sequentially and this "flag" might be send not last.

Reply:
You bring up a good point – I am ashamed I didn't think of this!

Reply:
Ok, I've fixed that problem now.